git_data_model.sql

```diff
@@ -1,160 +1,337 @@
-  repo_id BIGINT,
-  collaborator_id BIGINT,
-  FOREIGN KEY  (repo_id) REFERENCES users(user_id),
-  FOREIGN KEY  (collaborator_id) REFERENCES users(user_id)
+INSERT INTO repos (name,description,visibility)
+VALUES ('repo2','second repo of sal', 'public');
+
+SELECT * FROM repos;
+
+CREATE TABLE user_repos(
+  repo_id BIGINT NOT NULL,
+  user_id BIGINT NOT NULL,
+  access_level VARCHAR(255) NOT NULL,
+  FOREIGN KEY  (repo_id) REFERENCES repos(id),
+  FOREIGN KEY  (user_id) REFERENCES users(id)
 );
 
-INSERT INTO collaborators (repo_id,collaborator_id)
-VALUES (1,2);
+CREATE INDEX index_repo_access ON user_repos(repo_id,access_level);
+
+INSERT INTO user_repos (repo_id,user_id,access_level)
+VALUES (1,1,'owner');
+
 
-INSERT INTO collaborators (repo_id,collaborator_id)
-VALUES (1,3);
+INSERT INTO user_repos (repo_id,user_id,access_level)
+VALUES (1,2,'collaborator');
 
-SELECT * FROM collaborators;
+SELECT * FROM user_repos;
 
 
 CREATE TABLE commits(
-  commit_hash VARCHAR(255) PRIMARY KEY NOT NULL,
+  hash VARCHAR(255) PRIMARY KEY NOT NULL,
   parent_commit VARCHAR (255),
-  author BIGINT REFERENCES users(user_id),
+  user_id BIGINT NOT NULL REFERENCES users(id),
   messages TEXT,
-  dates DATE,
-  repo_id BIGINT,
-  FOREIGN KEY (repo_id) REFERENCES repo(repo_id),
-  FOREIGN KEY (parent_commit) REFERENCES commits(commit_hash),
-  UNIQUE (commit_hash)
+  created_at TIMESTAMP WITHOUT TIME ZONE DEFAULT(NOW() AT TIME ZONE 'UTC'),
+  repo_id BIGINT NOT NULL,
+  FOREIGN KEY (repo_id) REFERENCES repos(id),
+  FOREIGN KEY (parent_commit) REFERENCES commits(hash),
+  UNIQUE (hash)
 );
 
-INSERT INTO commits (commit_hash, author,messages,dates,repo_id)
-VALUES ('abcd', 1,'message1', '2022-01-01',1);
+CREATE INDEX index_parent_commit ON commits(parent_commit);
+CREATE INDEX index_commit_repo ON commits(repo_id);
+
+INSERT INTO commits (hash, user_id,messages,repo_id)
+VALUES ('abcd', 1,'message1',1);
+
+INSERT INTO commits (hash, parent_commit ,user_id,messages,repo_id)
+VALUES ('efgh','abcd', 1,'message2',1);
+
+INSERT INTO commits (hash,parent_commit, user_id,messages,repo_id)
+VALUES ('ijkl','efgh', 1,'message3',1);
 
-INSERT INTO commits (commit_hash, parent_commit ,author,messages,dates,repo_id)
-VALUES ('efgh','abcd', 1,'message2', '2022-02-01',1);
+INSERT INTO commits (hash,parent_commit, user_id,messages,repo_id)
+VALUES ('mnop','ijkl', 1,'message4',1);
 
-INSERT INTO commits (commit_hash,parent_commit, author,messages,dates,repo_id)
-VALUES ('ijkl','efgh', 1,'message3', '2022-03-01',1);
+INSERT INTO commits (hash,parent_commit, user_id,messages,repo_id)
+VALUES ('pqrs','mnop', 1,'message5',1);
 
+INSERT INTO commits (hash, user_id,messages,repo_id)
+VALUES ('aaa1', 2,'message5',2);
 
 SELECT * FROM commits;
 
 
 CREATE TABLE branches(
-  branch_id BIGSERIAL PRIMARY KEY NOT NULL,
-  branch_name VARCHAR (255),
-  head_pointer VARCHAR(255) ,
+  id BIGSERIAL PRIMARY KEY NOT NULL,
+  name VARCHAR (255) NOT NULL,
+  head_pointer VARCHAR(255) NOT NULL,
```

Comment:
what is head pointer?

Reply:
head pointer references to the latest commit hash